---
title: "Identification of patient-specific T cell neoantigens through HLA-agnostic genetic screens (Benchmark-CD8)"
author: "Thomas W. Battaglia"
date: "last update: `r format(Sys.Date(), format = '%d %B %Y')`"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
  theme: paper
  highlight: haddock
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(fig.path = "images/")
library(knitr)
library(patchwork)
library(ggsci)
library(DESeq2)
library(readxl)
library(tidyverse)
library(ggpubr)

# Function for calculating geometric mean
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
```

# Introduction
Reads were first de-multiplexed with ea-utils (fastq-multx) with an allowed barcode mismatch of 1 (Default). De-multiplexed reads were then subjected to an alignment against the UniVec vector contamination database using the adapter removal tool within ea-utils (fastq-mcf). Barcodes were extracted by remove the last 12 nucleotides from each read.

Next reads were subjected to an alignment against the Oligo library using a DNA aligner. We used Bowtie2...

## Data import & preprocessing
We are going to first import the merged counts tables that is outputted from the Nextflow pipeline. This table already has the raw counts summarised from the alignments and merged into a single matrix.


### Import oligos
-   oligo\_1899 (MART1\_ELA)
-   oligo\_4281 (MART1\_ELA)
-   oligo\_912 (MART1)
-   oligo\_3294 (MART1)
-   oligo\_1887 (CDK4mut)
-   oligo\_4269 (CDK4mut)
-   oligo\_1888 (CDK4wt)
-   oligo\_4270 (CDK4wt)
-   oligo\_1889 (RPS3Amut)
-   oligo\_4271 (RPS3Amut)
-   oligo\_1890 (RPS3Awt)
-   oligo\_4272 (RPS3Awt)
-   oligo\_1895 (MANSC1mut)
-   oligo\_4277 (MANSC1mut)
-   oligo\_1896 (MANSC1wt)
-   oligo\_4278 (MANSC1wt)

```{r}
# Import counts
cd8.counts = readRDS('../data/CD8-counts.rda')

# Convert to long format
cd8.counts.long = cd8.counts %>% 
  rownames_to_column("Id") %>% 
  pivot_longer(cols = -c('Id'), names_to = 'sample', values_to = 'counts')

# Import CD4 metadata
cd8.metadata = readRDS('../data/CD8-metadata.rda')

# Import MultiQC results
cd8.report = readRDS('../data/CD8-multiqc.rda')

# Import oligo annotations
oligo.genes = readRDS('../data/oligo-genes.rda')
```


### Summary stats

```{r}
# Mean
mean(cd8.report$raw_total_sequences)

# Median
median(cd8.report$raw_total_sequences)

# Range
range(cd8.report$raw_total_sequences)
```

### Mapped reads per sample

```{r}
cd8.report %>% 
  select(Sample, reads_unmapped, reads_mapped) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped", "reads_mapped"), names_to = "feature", values_to = "reads") %>% 
  ggplot(aes(x = Sample, y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_light(base_size = 11) +
  theme(legend.position = "top") +
  scale_fill_aaas() +
  xlab("") +
  ylab("Number of reads")
```

```{r}
cd8.report %>% 
  select(Sample, reads_unmapped_percent, reads_mapped_percent) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped_percent", "reads_mapped_percent"), names_to = "feature", values_to = "reads") %>% 
  mutate(reads = reads / 100) %>% 
  ggplot(aes(x = Sample, y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_light(base_size = 11) +
  theme(legend.position = "top") +
  scale_fill_aaas() +
  scale_y_continuous(labels = scales::percent) +
  xlab("") +
  ylab("Percent of reads")
```

### Mock replicates

```{r}
cd8.counts %>% 
  ggplot(., aes(x = log10(mock_a), y = log10(mock_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("Mock replicate #1 (log10)") +
  ylab("Mock replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```

### 1D3 replicates
```{r}
cd8.counts %>% 
  ggplot(aes(x = log10(id3_a), y = log10(id3_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1D3 replicate #1 (log10)") +
  ylab("1D3 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```


### CDK4-53 replicates

```{r}
cd8.counts %>% 
  ggplot(aes(x = log10(cdk453_a), y = log10(cdk453_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("CDK4/53 replicate #1 (log10)") +
  ylab("CDK4/53 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```


### Dilution 1:10 replicates

```{r}
cd8.counts %>% 
  ggplot(aes(x = log10(d1_10_a), y = log10(d1_10_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:10 replicate #1 (log10)") +
  ylab("1:10 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..))

```

### Dilution 1:100 replicates

```{r}
cd8.counts %>% 
  ggplot(aes(x = log10(d1_100_a), y = log10(d1_100_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:100 replicate #1 (log10)") +
  ylab("1:100 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```

### Dilution 1:300 replicates

```{r}
cd8.counts %>% 
  ggplot(aes(x = log10(d1_300_a), y = log10(d1_300_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:300 replicate #1 (log10)") +
  ylab("1:300 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```


### Dilution 1:1000 replicates

```{r}
cd8.counts %>% 
  ggplot(aes(x = log10(d1_1000_a), y = log10(d1_1000_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:1000 replicate #1 (log10)") +
  ylab("1:1000 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```

### Calculate oligo-rep variance
```{r}
# Calculate the variance between oligo-replicates
oligo.variance = cd8.counts %>% 
  rownames_to_column("Id") %>%
  mutate_if(is.numeric, funs(log10)) %>% 
  left_join(oligo.genes, by = "Id") %>% 
  group_by(oligoGroup) %>% 
  summarise(mock_a = diff(mock_a),
            mock_b = diff(mock_b),
            cdk453_a = diff(cdk453_a),
            cdk453_b = diff(cdk453_b),
            id3_a = diff(id3_a),
            id3_b = diff(id3_b),
            d1_10_a = diff(d1_10_a),
            d1_10_b = diff(d1_10_b),
            d1_100_a = diff(d1_100_a),
            d1_100_b = diff(d1_100_b),
            d1_300_a = diff(d1_300_a),
            d1_300_b = diff(d1_300_b),
            d1_1000_a = diff(d1_1000_a),
            d1_1000_b = diff(d1_1000_b))

# Plot histogram of variance across samples
oligo.variance %>% 
  pivot_longer(-oligoGroup, names_to = "Sample", values_to = "difference") %>% 
  ggplot(aes(x = difference)) +
  geom_histogram(bins = 100) +
  theme_bw() +
  facet_wrap(.~Sample) +
  xlab("Difference (log10)")
```

## Median-normalization

### Prepare data for DESeq2 object
```{r}
# Match the sample ID's between samples
cd8.counts = cd8.counts[,match(cd8.metadata$Sample, colnames(cd8.counts))]

# Make DEseq2 object
ddsMat.cd8 <- DESeqDataSetFromMatrix(countData = cd8.counts,
                                     colData = cd8.metadata,
                                     design = ~Group)
```


### Normalize 
```{r}
# Get normalize counts
ddsMat.cd8 <- estimateSizeFactors(ddsMat.cd8)
ddsMat.cd8 <- estimateDispersions(ddsMat.cd8, fitType='local')

# Get normalized counts 
cd8.norm <- counts(ddsMat.cd8, normalized = TRUE) %>% 
  as.data.frame() %>% 
  rownames_to_column("Id")

# Write tables to disk
write_tsv(cd8.counts, "tables/cd8-counts.tsv")
write_tsv(cd8.norm, "tables/cd8-counts-norm.tsv")

# Create a long table 
cd8.norm.long = cd8.norm %>% 
  gather(sample, norm.counts, -Id) 
```


### Plot boxplots
```{r}
# Pre-normalized
cd8.counts.long %>% 
  ggplot(aes(x = sample, y = log10(counts + 1))) +
  geom_boxplot() +
  theme_light(base_size = 11) +
  theme(axis.text.x = element_text(angle = 90)) +
  xlab("") +
  ylab("Counts (log10)") +
  ggtitle('Un-normalized data')

# After normalized
cd8.norm.long %>% 
  ggplot(aes(x = sample, y = log10(norm.counts + 1))) +
  geom_boxplot() +
  theme_light(base_size = 11) +
  theme(axis.text.x = element_text(angle = 90)) +
  xlab("") +
  ylab("Normalized counts (log10)") +
  ggtitle('Median-normalized data')
```

### Plot normalized histogram
```{r}
cd8.norm.long %>% 
  ggplot(aes(x = log10(norm.counts))) +
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  facet_wrap(sample~.) +
  theme_light(base_size = 11) +
  ylab("Density") +
  xlab("Normalized reads per oligo (log10)") 
```


## MART1/1D3 dropout

Oligo Id's for the positive control dropout's are: - oligo\_1899 (MART1\_ELA) - oligo\_4281 (MART1\_ELA) - oligo\_912 (MART1) - oligo\_3294 (MART1)

```{r}
# Get average between the replicates
id3.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         ID3_mean = rowMeans(select(., id3_a, id3_b))) %>% 
  mutate(M = log2(ID3_mean / Mock_mean),
         A = (log2(ID3_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
id3.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(ID3_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(id3.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(id3.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(id3.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(id3.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("1D3 normalized counts (log10)")

# Plot MA plot
id3.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(id3.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(id3.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(id3.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(id3.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " 1D3/Mock)"))) 
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("mock_a", "mock_b", "id3_a", "id3_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (ID3 / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
id3.mock.res = results(ddsMat.subset, alpha = 0.05)

# View table
id3.mock.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05)

# Show quick plot of significance ()
DESeq2::plotMA(id3.mock.res, ylim = c(-3, 3))
```

## CDK4/53 dropout

Oligo Id's for the positive control dropout's are: - oligo\_1899 (MART1\_ELA) - oligo\_4281 (MART1\_ELA) - oligo\_912 (MART1) - oligo\_3294 (MART1)

```{r}
# Get average between the replicates
cdk4.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         CDK4_mean = rowMeans(select(., cdk453_a, cdk453_b))) %>% 
  mutate(M = log2(CDK4_mean / Mock_mean),
         A = (log2(CDK4_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
cdk4.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(CDK4_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("CDK4 normalized counts (log10)")

# Plot MA plot
cdk4.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(cdk4.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " CDK4/Mock)"))) 
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("mock_a", "mock_b", "cdk453_a", "cdk453_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (CDK4 / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
cdk4.mock.res = results(ddsMat.subset, alpha = 0.05)

# View table
cdk4.mock.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05)

# Show quick plot of significance ()
DESeq2::plotMA(cdk4.mock.res)
```

## MART1/CDK4 dropout

Oligo Id's for the positive control dropout's are: - oligo\_1899 (MART1\_ELA) - oligo\_4281 (MART1\_ELA) - oligo\_912 (MART1) - oligo\_3294 (MART1)

```{r}
# Get average between the replicates
id3.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         CDK4_mean = rowMeans(select(., cdk453_a, cdk453_b)),
         ID3_mean = rowMeans(select(., id3_a, id3_b))) %>% 
  mutate(M = log2(ID3_mean / CDK4_mean),
         A = (log2(ID3_mean) + log2(CDK4_mean)) / 2) 

# Plot scatterplot
id3.mean %>% 
  ggplot(aes(x = log10(CDK4_mean), y = log10(ID3_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(id3.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(id3.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(id3.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(id3.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("CDK4 normalized counts (log10)") +
  ylab("1D3 normalized counts (log10)") 

# Plot MA plot
id3.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(id3.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(id3.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(id3.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(id3.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " 1D3/CDK4)"))) 
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("cdk453_a", "cdk453_b", "id3_a", "id3_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "CDK4")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (ID3 / CDK4)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
id3.cdk4.res = results(ddsMat.subset, alpha = 0.05)

# View table
id3.cdk4.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05 & abs(log2FoldChange) > 1)

# Show quick plot of significance ()
DESeq2::plotMA(id3.cdk4.res)
```



## Dilutions 1:10 dropout


```{r}
# Get average between the replicates
d10.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D10_mean = rowMeans(select(., d1_10_a, d1_10_b))) %>% 
  mutate(M = log2(D10_mean / Mock_mean),
         A = (log2(D10_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
d10.scatter = d10.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(D10_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d10.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d10.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d10.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d10.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("D10 normalized counts (log10)")
d10.scatter

# Plot MA plot
d10.ma = d10.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d10.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d10.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d10.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d10.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  ylim(-5, 5) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " D10/Mock)")))
d10.ma
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("mock_a", "mock_b", "d1_10_a", "d1_10_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (D10 / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
d10.mock.res = results(ddsMat.subset, alpha = 0.05)

# View table
d10.mock.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05)

# Show quick plot of significance ()
DESeq2::plotMA(d10.mock.res & abs(log2FoldChange) > 1)
```


## Dilutions 1:100 dropout

```{r}
# Get average between the replicates
d100.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D100_mean = rowMeans(select(., d1_100_a, d1_100_b))) %>% 
  mutate(M = log2(D100_mean / Mock_mean),
         A = (log2(D100_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
d100.scatter = d100.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(D100_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d100.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d100.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d100.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d100.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("D100 normalized counts (log10)")
d100.scatter

# Plot MA plot
d100.ma = d100.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d100.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d100.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d100.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d100.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  ylim(-5, 5) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " D100/Mock)")))
d100.ma
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("mock_a", "mock_b", "d1_100_a", "d1_100_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (D100 / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
d100.mock.res = results(ddsMat.subset, alpha = 0.05)

# View table
d100.mock.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05 & abs(log2FoldChange) > 1)

# Show quick plot of significance ()
DESeq2::plotMA(d100.mock.res)
```


## Dilutions 1:300 dropout

```{r}
# Get average between the replicates
d300.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D300_mean = rowMeans(select(., d1_300_a, d1_300_b))) %>% 
  mutate(M = log2(D300_mean / Mock_mean),
         A = (log2(D300_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
d300.scatter = d300.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(D300_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d300.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d300.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d300.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d300.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("D300 normalized counts (log10)")
d300.scatter

# Plot MA plot
d300.ma = d300.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d300.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d300.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d300.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d300.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  ylim(-5, 5) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " D300/Mock)")))
d300.ma
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("mock_a", "mock_b", "d1_300_a", "d1_300_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (D300 / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
d300.mock.res = results(ddsMat.subset, alpha = 0.05)

# View table
d300.mock.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05 & abs(log2FoldChange) > 1)

# Show quick plot of significance ()
DESeq2::plotMA(d300.mock.res)
```


## Dilutions 1:1000 dropout

```{r}
# Get average between the replicates
d1000.mean = cd8.norm %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D1000_mean = rowMeans(select(., d1_1000_a, d1_1000_b))) %>% 
  mutate(M = log2(D1000_mean / Mock_mean),
         A = (log2(D1000_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
d1000.scatter = d1000.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(D1000_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d1000.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d1000.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d1000.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d1000.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("D1000 normalized counts (log10)")
d1000.scatter

# Plot MA plot
d1000.ma = d1000.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = alpha("red", 0.7), data = filter(d1000.mean, Id %in% c("oligo_4281", "oligo_1899"))) +
  geom_point(colour = alpha("orange", 0.7), data = filter(d1000.mean, Id %in% c("oligo_1888", "oligo_4270"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = filter(d1000.mean, Id %in% c("oligo_913", "oligo_3295"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = filter(d1000.mean, Id %in% c("oligo_1887", "oligo_4269"))) +
  theme_light(base_size = 11) +
  ylim(-5, 5) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " D1000/Mock)")))
d1000.ma
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd8[,which(colnames(assay(ddsMat.cd8)) %in% c("mock_a", "mock_b", "d1_1000_a", "d1_1000_b"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset)

# Get results (D1000 / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
d1000.mock.res = results(ddsMat.subset, alpha = 0.05)

# View table
d1000.mock.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>% 
  filter(padj < 0.05)

# Show quick plot of significance ()
DESeq2::plotMA(d1000.mock.res)
```


## Plot serial dilutions together
```{r}
# Scatter plot
d10.scatter + d100.scatter + d300.scatter + d1000.scatter 

# MA plot
d10.ma + d100.ma + d300.ma + d1000.ma
```



## Filter low count oligos (visualization)
```{r}
# Rank
cd8.norm.long %>% 
  group_by(Id) %>% 
  summarise(sum = sum(norm.counts),
            mean = mean(norm.counts),
            median = median(norm.counts),
            sd = sd(norm.counts),
            geomean = gm_mean(norm.counts),
            cv = sd(norm.counts) / mean(norm.counts))  %>% 
  mutate(rank = rank(mean)) %>% 
  ggplot(aes(x = rank, y = log10(mean))) +
  geom_point() +
  theme_minimal()
```

### Plot average abundance 
```{r}
# Get mean counts per oligo
cd8.mean = cd8.norm.long %>% 
  group_by(Id) %>% 
  summarise(sum = sum(norm.counts),
            mean = mean(norm.counts),
            median = median(norm.counts),
            sd = sd(norm.counts),
            geomean = gm_mean(norm.counts),
            cv = sd(norm.counts) / mean(norm.counts)) 

# Get quantiles
quantile(cd8.mean$median , seq(from = 0, to = 0.10, by = 0.005))

# Histogram with 5% cutoff
cd8.norm.long %>% 
  group_by(Id) %>% 
  summarise(total.count = sum(norm.counts),
            mean.count = mean(norm.counts),
            median = median(norm.counts),
            geomean = gm_mean(norm.counts),
            sd.count = sd(norm.counts)) %>% 
  ggplot(aes(x = log10(median))) +
  geom_histogram(binwidth = 0.25, colour = "black", fill = "white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  geom_vline(xintercept = log10(396.491177), color = "red") +
  theme_light(base_size = 11) +
  xlab("Normalized reads per oligo (log10)")
```

### Apply 5% cutoff
```{r}
# Get ID's within the limits
## Cutoff = 5%
idx <- cd8.norm.long %>% 
  group_by(Id) %>% 
  summarise(sum = sum(norm.counts),
            mean = mean(norm.counts),
            median = median(norm.counts),
            geomean = gm_mean(norm.counts),
            sd = sd(norm.counts) ,
            c.v = sd/mean) %>% 
  filter(median >= 396.491177) %>% 
  pull(Id)

# Filter table
cd8.norm.fil = cd8.norm %>% 
  filter(Id %in% idx)

# Write filter table to disk
write_tsv(cd8.norm, "tables/cd8-counts-norm-filtered.tsv")

# Get dimensions
dim(cd8.norm)
dim(cd8.norm.fil)

# Create long dataframe
cd8.norm.fil.long = cd8.norm.fil %>% 
  gather(sample, norm.counts, -Id) 

# Plot boxplot
cd8.norm.fil.long %>% 
  ggplot(aes(x = sample, y = log10(norm.counts + 1))) +
  geom_boxplot() +
  theme_light(base_size = 11) +
  theme(axis.text.x = element_text(angle = 90)) +
  xlab("") +
  ylab("Normalized counts (log10)") +
  ggtitle('Median-normalized data w/ filtering') 

# Histogram all
cd8.norm.fil.long %>% 
  ggplot(aes(x = log10(norm.counts))) +
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  facet_wrap(sample~.) +
  theme_light(base_size = 11) +
  ylab("Density") +
  xlab("(Filtered) Normalized reads per oligo (log10)")
```

### MA plot (1D3/CDK4.53)
```{r}
# Get average between the replicates
id3.mean = cd8.norm.fil %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         CDK4_mean = rowMeans(select(., cdk453_a, cdk453_b)),
         ID3_mean = rowMeans(select(., id3_a, id3_b))) %>% 
  mutate(M = log2(ID3_mean / CDK4_mean),
         A = (log2(ID3_mean) + log2(CDK4_mean)) / 2) 

# Plot MA plot
id3.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(size = 3.5, colour = rgb(237,157,148, max=255), data = filter(id3.mean, Id %in% c("oligo_4281", "oligo_1899"))) + #MART1-Ag
  geom_point(size = 3.5, colour = rgb(199,150,149, max=255), data = filter(id3.mean, Id %in% c("oligo_913", "oligo_3295"))) + #MART1-W
  geom_point(size = 3.5, colour = rgb(102,165,165, max=255), data = filter(id3.mean, Id %in% c("oligo_1887", "oligo_4269"))) + #CDK4-Ag
  geom_point(size = 3.5, colour = rgb(88,121,163, max=255), data = filter(id3.mean, Id %in% c("oligo_1888", "oligo_4270"))) + #CDK4-WT
  theme_light(base_size = 11) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " 1D3/CDK4)"))) 
```
### Titration data
```{r}
# 1:10
d10.mean = cd8.norm.fil %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D10_mean = rowMeans(select(., d1_10_a, d1_10_b))) %>% 
  mutate(M = log2(D10_mean / Mock_mean),
         A = (log2(D10_mean) + log2(Mock_mean)) / 2) 

d10.plot = d10.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(size = 4.5, colour = rgb(237,157,148, max=255), data = filter(d10.mean, Id %in% c("oligo_4281", "oligo_1899"))) + #MART1-Ag
  geom_point(size = 4.5, colour = rgb(199,150,149, max=255), data = filter(d10.mean, Id %in% c("oligo_913", "oligo_3295"))) + #MART1-W
  geom_point(size = 4.5, colour = rgb(102,165,165, max=255), data = filter(d10.mean, Id %in% c("oligo_1887", "oligo_4269"))) + #CDK4-Ag
  geom_point(size = 4.5, colour = rgb(88,121,163, max=255), data = filter(d10.mean, Id %in% c("oligo_1888", "oligo_4270"))) + #CDK4-WT
  theme_light(base_size = 11) +
  ylim(-5, 1) +
  xlim(7, 14) +
  xlab("") +
  ylab("")
d10.plot 

# 1:100
d100.mean = cd8.norm.fil %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D100_mean = rowMeans(select(., d1_100_a, d1_100_b))) %>% 
  mutate(M = log2(D100_mean / Mock_mean),
         A = (log2(D100_mean) + log2(Mock_mean)) / 2) 

d100.plot = d100.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(size = 4.5, colour = rgb(237,157,148, max=255), data = filter(d100.mean, Id %in% c("oligo_4281", "oligo_1899"))) + #MART1-Ag
  geom_point(size = 4.5, colour = rgb(199,150,149, max=255), data = filter(d100.mean, Id %in% c("oligo_913", "oligo_3295"))) + #MART1-W
  geom_point(size = 4.5, colour = rgb(102,165,165, max=255), data = filter(d100.mean, Id %in% c("oligo_1887", "oligo_4269"))) + #CDK4-Ag
  geom_point(size = 4.5, colour = rgb(88,121,163, max=255), data = filter(d100.mean, Id %in% c("oligo_1888", "oligo_4270"))) + #CDK4-WT
  theme_light(base_size = 11) +
  ylim(-5, 1) +
  xlim(7, 14) +
  xlab("") +
  ylab("")
d100.plot

# 1:300
d300.mean = cd8.norm.fil %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D300_mean = rowMeans(select(., d1_300_a, d1_300_b))) %>% 
  mutate(M = log2(D300_mean / Mock_mean),
         A = (log2(D300_mean) + log2(Mock_mean)) / 2) 

d300.plot = d300.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(size = 4.5, colour = rgb(237,157,148, max=255), data = filter(d300.mean, Id %in% c("oligo_4281", "oligo_1899"))) + #MART1-Ag
  geom_point(size = 4.5, colour = rgb(199,150,149, max=255), data = filter(d300.mean, Id %in% c("oligo_913", "oligo_3295"))) + #MART1-W
  geom_point(size = 4.5, colour = rgb(102,165,165, max=255), data = filter(d300.mean, Id %in% c("oligo_1887", "oligo_4269"))) + #CDK4-Ag
  geom_point(size = 4.5, colour = rgb(88,121,163, max=255), data = filter(d300.mean, Id %in% c("oligo_1888", "oligo_4270"))) + #CDK4-WT
  theme_light(base_size = 11) +
  ylim(-5, 1) +
  xlim(7, 14) +
  xlab("") +
  ylab(expression(paste("Fold change ", '(', log[2], " D300/Mock)"))) 
d300.plot

# 1:1000
d1000.mean = cd8.norm.fil %>% 
  mutate(Mock_mean = rowMeans(select(., mock_a, mock_b)),
         D1000_mean = rowMeans(select(., d1_1000_a, d1_1000_b))) %>% 
  mutate(M = log2(D1000_mean / Mock_mean),
         A = (log2(D1000_mean) + log2(Mock_mean)) / 2) 

d1000.plot = d1000.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(size = 4.5, colour = rgb(237,157,148, max=255), data = filter(d1000.mean, Id %in% c("oligo_4281", "oligo_1899"))) + #MART1-Ag
  geom_point(size = 4.5, colour = rgb(199,150,149, max=255), data = filter(d1000.mean, Id %in% c("oligo_913", "oligo_3295"))) + #MART1-W
  geom_point(size = 4.5, colour = rgb(102,165,165, max=255), data = filter(d1000.mean, Id %in% c("oligo_1887", "oligo_4269"))) + #CDK4-Ag
  geom_point(size = 4.5, colour = rgb(88,121,163, max=255), data = filter(d1000.mean, Id %in% c("oligo_1888", "oligo_4270"))) + #CDK4-WT
  theme_light(base_size = 11) +
  ylim(-5, 1) +
  xlim(7, 14) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab("")
d1000.plot

# Plot as single panel
d10.plot / d100.plot / d300.plot / d1000.plot
```


```{r}
sessionInfo()
```


