---
title: "Identification of patient-specific T cell neoantigens through HLA-agnostic genetic screens (Benchmark-CD4)"
author: "Thomas W. Battaglia"
date: "last update: `r format(Sys.Date(), format = '%d %B %Y')`"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
  theme: paper
  highlight: haddock
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(fig.path = "images/")
library(knitr)
library(patchwork)
library(ggsci)
library(DESeq2)
library(readxl)
library(tidyverse)
library(ggpubr)

# Function for calculating geometric mean
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
```

# Introduction
Reads were first de-multiplexed with ea-utils (fastq-multx) with an allowed barcode mismatch of 1 (Default). De-multiplexed reads were then subjected to an alignment against the UniVec vector contamination database using the adapter removal tool within ea-utils (fastq-mcf). Barcodes were extracted by remove the last 12 nucleotides from each read.

Next reads were subjected to an alignment against the Oligo library using a DNA aligner. We used Bowtie2...

## Data import & preprocessing
We are going to first import the merged counts tables that is outputted from the Nextflow pipeline. This table already has the raw counts summarised from the alignments and merged into a single matrix.

```{r}
# Import counts
cd4.counts = readRDS('../data/CD4-counts.rda') %>% 
  rownames_to_column("Id") %>% 
  mutate(Id = recode(Id, oligo_1899 = "MART1_ELA_a",
                         oligo_4281 = "MART1_ELA_b",
                         oligo_913 = "MART1_a",
                         oligo_3295 = "MART1_b",
                         oligo_1887 = "CDK4mut_a",
                         oligo_4269 = "CDK4mut_b",
                         oligo_1888 = "CDK4wt_a",
                         oligo_4270 = "CDK4wt_b",
                         oligo_1889 = "RPS3Amut_a",
                         oligo_4271 = "RPS3Amut_b",
                         oligo_1890 = "RPS3Awt_a",
                         oligo_4272 = "RPS3Awt_b",
                         oligo_1895 = "MANSC1mut_a",
                         oligo_4277 = "MANSC1mut_b",
                         oligo_1896 = "MANSC1wt_a",
                         oligo_4278 = "MANSC1wt_b")) %>% 
  column_to_rownames("Id")

# Convert to long format
cd4.counts.long = cd4.counts %>% 
  rownames_to_column("Id") %>% 
  pivot_longer(cols = starts_with('cd4'), names_to = 'sample', values_to = 'counts')

# Import CD4 metadata
cd4.metadata = readRDS('../data/CD4-metadata.rda')

# Import MultiQC results
cd4.report = readRDS('../data/CD4-multiqc.rda')

# Import oligo annotations
oligo.genes = readRDS('../data/oligo-genes.rda')
```


### Summary stats

```{r}
# Mean
mean(cd4.report$raw_total_sequences)

# Median
median(cd4.report$raw_total_sequences)

# Range
range(cd4.report$raw_total_sequences)
```

### Mapped reads per sample

```{r}
cd4.report %>% 
  select(Sample, reads_unmapped, reads_mapped) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped", "reads_mapped"), names_to = "feature", values_to = "reads") %>% 
  ggplot(aes(x = Sample, y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_light(base_size = 11) +
  theme(legend.position = "top") +
  scale_fill_aaas() +
  xlab("") +
  ylab("Number of reads")
```

```{r}
cd4.report %>% 
  select(Sample, reads_unmapped_percent, reads_mapped_percent) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped_percent", "reads_mapped_percent"), names_to = "feature", values_to = "reads") %>% 
  mutate(reads = reads / 100) %>% 
  ggplot(aes(x = Sample, y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_light(base_size = 11) +
  theme(legend.position = "top") +
  scale_fill_aaas() +
  scale_y_continuous(labels = scales::percent) +
  xlab("") +
  ylab("Percent of reads")
```

### Mock replicates
```{r}
cd4.counts %>% 
  ggplot(aes(x = log10(cd4_mock_1), y = log10(cd4_mock_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("Mock replicate #1 (log10)") +
  ylab("Mock replicate #2 (log10)") +
  theme_light() +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE)
```
### SNORD73 replicates

```{r}
cd4.counts %>% 
  ggplot(aes(x = log10(cd4_snord73_1), y = log10(cd4_snord73_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("SNORD73 replicate #1 (log10)") +
  ylab("SNORD73 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE)

cd4.counts %>% 
  ggplot(aes(x = log10(cd4_snord73_2), y = log10(cd4_snord73_3))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("SNORD73 replicate #1 (log10)") +
  ylab("SNORD73 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 

cd4.counts %>% 
  ggplot(aes(x = log10(cd4_snord73_1), y = log10(cd4_snord73_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("SNORD73 replicate #1 (log10)") +
  ylab("SNORD73 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE)
```


### MANSC1 replicates

```{r}
cd4.counts %>% 
  ggplot(aes(x = log10(cd4_mansc1_1), y = log10(cd4_mansc1_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("MANSC1 replicate #1 (log10)") +
  ylab("MANSC1 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 

cd4.counts %>% 
  ggplot(aes(x = log10(cd4_mansc1_2), y = log10(cd4_mansc1_3))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("MANSC1 replicate #2 (log10)") +
  ylab("MANSC1 replicate #3 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 

cd4.counts %>% 
  ggplot(aes(x = log10(cd4_mansc1_1), y = log10(cd4_mansc1_3))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("MANSC1 replicate #1 (log10)") +
  ylab("MANSC1 replicate #3 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 
```



## Median-normalization

### Prepare data for DESeq2 object
```{r}
# Match the sample ID's between samples
cd4.counts = cd4.counts[,match(cd4.metadata$Sample, colnames(cd4.counts))]

# Make DEseq2 object
ddsMat.cd4 <- DESeqDataSetFromMatrix(countData = cd4.counts,
                                     colData = cd4.metadata,
                                     design = ~Group)
```


### Normalize 

```{r}
# Get normalize counts
ddsMat.cd4 <- estimateSizeFactors(ddsMat.cd4)
ddsMat.cd4 <- estimateDispersions(ddsMat.cd4, fitType='local')

# Get normalized counts
cd4.norm <- counts(ddsMat.cd4, normalized = TRUE) %>% 
  as.data.frame() %>% 
  rownames_to_column("Id")

# Write tables to disk
write_tsv(cd4.counts, "tables/cd4-counts.tsv")
write_tsv(cd4.norm, "tables/cd4-counts-norm.tsv")

# Create a long table 
cd4.norm.long = cd4.norm %>% 
  gather(sample, norm.counts, -Id) 
```


### Export data for Mageck
```{r}
cd4.counts %>% 
  rownames_to_column("Id") %>% 
  left_join(oligo.genes) %>% 
  select(Id, `GENE NAME`, everything(), -oligoGroup) %>% 
  dplyr::rename("sgDNA" = "Id",
                "Gene" = "GENE NAME") %>% 
  write.table("tables/mageck_counts-cd4.txt", sep = "\t", quote = F, row.names = F) 
```

### Plot boxplots
```{r}
# Pre-normalized
cd4.counts.long %>% 
  ggplot(aes(x = sample, y = log10(counts + 1))) +
  geom_boxplot() +
  theme_light(base_size = 11) +
  theme(axis.text.x = element_text(angle = 90)) +
  xlab("") +
  ylab("Counts (log10)") +
  ggtitle('Un-normalized data')

# After normalized
cd4.norm.long %>% 
  ggplot(aes(x = sample, y = log10(norm.counts + 1))) +
  geom_boxplot() +
  theme_light(base_size = 11) +
  theme(axis.text.x = element_text(angle = 90)) +
  xlab("") +
  ylab("Normalized counts (log10)") +
  ggtitle('Median-normalized data')
```

### Plot normalized histogram
```{r}
cd4.norm.long %>% 
  ggplot(aes(x = log10(norm.counts))) +
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  facet_wrap(sample~.) +
  theme_light(base_size = 11) +
  ylab("Density") +
  xlab("Normalized reads per oligo (log10)") 
```

## MANSC1 dropout
```{r}
# Get average between the replicates
cd4.mansc1.mean = cd4.norm %>% 
  mutate(Mock_mean = rowMeans(select(., cd4_mock_1, cd4_mock_2)),
         MANSC1_mean = rowMeans(select(., cd4_mansc1_1, cd4_mansc1_2, cd4_mansc1_3))) %>% 
  mutate(M = log2(MANSC1_mean / Mock_mean),
         A = (log2(MANSC1_mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
p1 = cd4.mansc1.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(MANSC1_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = "red", data = . %>% filter(Id %in% c("MANSC1mut_a", "MANSC1mut_b"))) +
  geom_point(colour = "blue", data = . %>% filter(Id %in% c("MANSC1wt_a", "MANSC1wt_b"))) + 
  geom_point(colour = "darkgreen", data = . %>% filter(Id %in% c("RPS3Amut_a", "RPS3Amut_b"))) + 
  geom_point(colour = "orange", data = . %>% filter(Id %in% c("RPS3Awt_a", "RPS3Awt_b"))) + 
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("Mansc1 normalized counts (log10)") 

# Plot MA plot
p2 = cd4.mansc1.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = "red", data = . %>% filter(Id %in% c("MANSC1mut_a", "MANSC1mut_b"))) +
  geom_point(colour = "blue", data = . %>% filter(Id %in% c("MANSC1wt_a", "MANSC1wt_b"))) + 
  geom_point(colour = "darkgreen", data = . %>% filter(Id %in% c("RPS3Amut_a", "RPS3Amut_b"))) + 
  geom_point(colour = "orange", data = . %>% filter(Id %in% c("RPS3Awt_a", "RPS3Awt_b"))) + 
  theme_light(base_size = 11) +
  ylim(-3, 3) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " Mansc1/Mock)")))

p1 + p2
```
#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd4[,which(colnames(assay(ddsMat.cd4)) %in% c("cd4_mock_1", "cd4_mock_2", "cd4_mansc1_1", "cd4_mansc1_2", "cd4_mansc1_3"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset, fitType='local')

# Get results (MANSC / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
cd4.mansc1.res = results(ddsMat.subset, alpha = 0.05)

# View table
cd4.mansc1.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>%  
  filter(padj < 0.05, abs(log2FoldChange) > 1)

# Show quick plot of significance (Unknown oligo's)
DESeq2::plotMA(cd4.mansc1.res, ylim = c(-3,3))
```


## Snord73 dropout
```{r}
# Get average between the replicates
cd4.snord73.mean = cd4.norm %>% 
  mutate(Mock_mean = rowMeans(select(., cd4_mock_1, cd4_mock_2)),
         SNORD73mean = rowMeans(select(., cd4_snord73_1, cd4_snord73_2, cd4_snord73_3))) %>% 
  mutate(M = log2(SNORD73mean / Mock_mean),
         A = (log2(SNORD73mean) + log2(Mock_mean)) / 2) 

# Plot scatterplot
p1 = cd4.snord73.mean %>% 
  ggplot(aes(x = log10(Mock_mean), y = log10(SNORD73mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = "red", data = . %>% filter(Id %in% c("MANSC1mut_a", "MANSC1mut_b"))) +
  geom_point(colour = "blue", data = . %>% filter(Id %in% c("MANSC1wt_a", "MANSC1wt_b"))) + 
  geom_point(colour = "darkgreen", data = . %>% filter(Id %in% c("RPS3Amut_a", "RPS3Amut_b"))) + 
  geom_point(colour = "orange", data = . %>% filter(Id %in% c("RPS3Awt_a", "RPS3Awt_b"))) + 
  theme_light(base_size = 11) +
  xlab("Mock normalized counts (log10)") +
  ylab("Snord73a normalized counts (log10)") 

# Plot MA plot
p2 = cd4.snord73.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = "red", data = . %>% filter(Id %in% c("MANSC1mut_a", "MANSC1mut_b"))) +
  geom_point(colour = "blue", data = . %>% filter(Id %in% c("MANSC1wt_a", "MANSC1wt_b"))) + 
  geom_point(colour = "darkgreen", data = . %>% filter(Id %in% c("RPS3Amut_a", "RPS3Amut_b"))) + 
  geom_point(colour = "orange", data = . %>% filter(Id %in% c("RPS3Awt_a", "RPS3Awt_b"))) + 
  theme_light(base_size = 11) +
  ylim(-3, 3) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " Snord73a/Mock)")))

p1 + p2
```
#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd4[,which(colnames(assay(ddsMat.cd4)) %in% c("cd4_mock_1", "cd4_mock_2", "cd4_snord73_1", "cd4_snord73_2", "cd4_snord73_3"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "Mock")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset, fitType='local')

# Get results (MANSC / Mock)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
cd4.snord.res = results(ddsMat.subset, alpha = 0.05)

# View table
cd4.snord.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>%  
  filter(padj < 0.05, abs(log2FoldChange) > 1)

# Show quick plot of significance (Unknown oligo's)
DESeq2::plotMA(cd4.snord.res, ylim = c(-3,3))
```

## Snord73 vs. MANSC1 dropout

```{r}
# Get average between the replicates
cd4.mansc1.snord73.mean = cd4.norm %>% 
  mutate(SNORD73mean = rowMeans(select(., cd4_snord73_1, cd4_snord73_2, cd4_snord73_3)),
         Mock_mean = rowMeans(select(., cd4_mock_1, cd4_mock_2)),
         MANSC1_mean = rowMeans(select(., cd4_mansc1_1, cd4_mansc1_2, cd4_mansc1_3))) %>% 
  mutate(M = log2(MANSC1_mean / SNORD73mean),
         A = (log2(MANSC1_mean) + log2(SNORD73mean)) / 2) 

# Plot scatterplot
p1 = cd4.mansc1.snord73.mean %>% 
  ggplot(aes(x = log10(SNORD73mean), y = log10(MANSC1_mean))) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = "red", data = . %>% filter(Id %in% c("MANSC1mut_a", "MANSC1mut_b"))) +
  geom_point(colour = "blue", data = . %>% filter(Id %in% c("MANSC1wt_a", "MANSC1wt_b"))) + 
  geom_point(colour = "darkgreen", data = . %>% filter(Id %in% c("RPS3Amut_a", "RPS3Amut_b"))) + 
  geom_point(colour = "orange", data = . %>% filter(Id %in% c("RPS3Awt_a", "RPS3Awt_b"))) + 
  theme_light(base_size = 11) +
  xlab("Snord73a normalized counts (log10)") +
  ylab("Mansc1 normalized counts (log10)") 

# Plot MA plot
p2 = cd4.mansc1.snord73.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(colour = "red", data = . %>% filter(Id %in% c("MANSC1mut_a", "MANSC1mut_b"))) +
  geom_point(colour = "blue", data = . %>% filter(Id %in% c("MANSC1wt_a", "MANSC1wt_b"))) + 
  geom_point(colour = "darkgreen", data = . %>% filter(Id %in% c("RPS3Amut_a", "RPS3Amut_b"))) + 
  geom_point(colour = "orange", data = . %>% filter(Id %in% c("RPS3Awt_a", "RPS3Awt_b"))) + 
  theme_light(base_size = 11) +
  ylim(-3, 3) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " Mansc1/Snord73a)")))

p1 + p2
```

#### Calculate statistics 
```{r}
# Create new DESeq2 object with poor samples removed
ddsMat.subset = ddsMat.cd4[,which(colnames(assay(ddsMat.cd4)) %in% c("cd4_snord73_1", "cd4_snord73_2", "cd4_snord73_3", "cd4_mansc1_1", "cd4_mansc1_2", "cd4_mansc1_3"))]
ddsMat.subset$Group = droplevels(ddsMat.subset$Group)
ddsMat.subset$Group = relevel(ddsMat.subset$Group, "SNORD")

# Run stats
ddsMat.subset <- DESeq(ddsMat.subset, fitType='local')

# Get results (MANSC / SNORD)
results(ddsMat.subset, alpha = 0.05) %>% 
  summary()

# Get table of results
cd4.mansc.snord.res = results(ddsMat.subset, alpha = 0.05)

# View table
cd4.mansc.snord.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Id") %>%  
  filter(padj < 0.05, abs(log2FoldChange) > 1)

# Show quick plot of significance (Unknown oligo's)
DESeq2::plotMA(cd4.mansc.snord.res, ylim = c(-3,3))
```


## Filter low count oligos (visualization)
```{r}
# Rank
cd4.norm.long %>% 
  group_by(Id) %>% 
  summarise(sum = sum(norm.counts),
            mean = mean(norm.counts),
            median = median(norm.counts),
            sd = sd(norm.counts),
            geomean = gm_mean(norm.counts),
            cv = sd(norm.counts) / mean(norm.counts))  %>% 
  mutate(rank = rank(mean)) %>% 
  ggplot(aes(x = rank, y = log10(mean))) +
  geom_point() +
  theme_minimal()
```

### Plot average abundance 
```{r}
# Get mean counts per oligo
cd4.mean = cd4.norm.long %>% 
  group_by(Id) %>% 
  summarise(sum = sum(norm.counts),
            mean = mean(norm.counts),
            median = median(norm.counts),
            sd = sd(norm.counts),
            geomean = gm_mean(norm.counts),
            cv = sd(norm.counts) / mean(norm.counts)) 

# Get quantiles
quantile(cd4.mean$median , seq(from = 0, to = 0.10, by = 0.005))

# Histogram with 5% cutoff
cd4.norm.long %>% 
  group_by(Id) %>% 
  summarise(total.count = sum(norm.counts),
            mean.count = mean(norm.counts),
            median = median(norm.counts),
            geomean = gm_mean(norm.counts),
            sd.count = sd(norm.counts)) %>% 
  ggplot(aes(x = log10(median))) +
  geom_histogram(binwidth = 0.25, colour = "black", fill = "white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  geom_vline(xintercept = log10(591.123693), color = "red") +
  theme_light(base_size = 11) +
  xlab("Normalized reads per oligo (log10)")
```
### Apply 5% cutoff
```{r}
# Get ID's within the limits
## Cutoff = 5%
idx <- cd4.norm.long %>% 
  group_by(Id) %>% 
  summarise(sum = sum(norm.counts),
            mean = mean(norm.counts),
            median = median(norm.counts),
            geomean = gm_mean(norm.counts),
            sd = sd(norm.counts) ,
            c.v = sd/mean) %>% 
  filter(median >= 591.123693) %>% 
  pull(Id)

# Filter table
cd4.norm.fil = cd4.norm %>% 
  filter(Id %in% idx)

# Write filter table to disk
write_tsv(cd4.norm, "tables/cd4-counts-norm-filtered.tsv")

# Get dimensions
dim(cd4.norm)
dim(cd4.norm.fil)

# Create long dataframe
cd4.norm.fil.long = cd4.norm.fil %>% 
  gather(sample, norm.counts, -Id) 

# Plot boxplot
cd4.norm.fil.long %>% 
  ggplot(aes(x = sample, y = log10(norm.counts + 1))) +
  geom_boxplot() +
  theme_light(base_size = 11) +
  theme(axis.text.x = element_text(angle = 90)) +
  xlab("") +
  ylab("Normalized counts (log10)") +
  ggtitle('Median-normalized data w/ filtering') 

# Histogram all
cd4.norm.fil.long %>% 
  ggplot(aes(x = log10(norm.counts))) +
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  facet_wrap(sample~.) +
  theme_light(base_size = 11) +
  ylab("Density") +
  xlab("(Filtered) Normalized reads per oligo (log10)")
```

###  Scatter/MA plot (MANSC1/SNORD73)
```{r}
cd4.mansc1.snord73.mean =  cd4.norm.fil %>% 
  mutate(SNORD73mean = rowMeans(select(., cd4_snord73_1, cd4_snord73_2, cd4_snord73_3)),
         Mock_mean = rowMeans(select(., cd4_mock_1, cd4_mock_2)),
         MANSC1_mean = rowMeans(select(., cd4_mansc1_1, cd4_mansc1_2, cd4_mansc1_3))) %>% 
  mutate(M = log2(MANSC1_mean / SNORD73mean),
         A = (log2(MANSC1_mean) + log2(SNORD73mean)) / 2) 

cd4.mansc1.snord73.mean %>% 
  ggplot(aes(x = A, y = M)) +
  geom_point(colour = alpha("grey", 0.5)) +
  geom_point(size = 3.5, colour = rgb(152,179,136, max=255), data = filter(cd4.mansc1.snord73.mean, Id %in% c("oligo_1895", "oligo_4277"))) +
  geom_point(size = 3.5, colour = rgb(135,135,135, max=255), data = filter(cd4.mansc1.snord73.mean, Id %in% c("oligo_1896", "oligo_4278"))) +
  geom_point(size = 3.5, colour = rgb(233,175,107, max=255), data = filter(cd4.mansc1.snord73.mean, Id %in% c("oligo_1889", "oligo_4271"))) +
  geom_point(size = 3.5, colour = rgb(135,135,135, max=255), data = filter(cd4.mansc1.snord73.mean, Id %in% c("oligo_1890", "oligo_4272"))) +
  theme_light(base_size = 10.5) +
  ylim(-3, 3) +
  xlab(expression(paste("Mean normalized counts ", '(', log[2], ")"))) +
  ylab(expression(paste("Fold change ", '(', log[2], " Mansc1/Snord73A)"))) 
```



```{r}
sessionInfo()
```






