---
title: "CD4-Benchmark"
author: "Thomas W. Battaglia"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(fig.path = "images/")
library(knitr)
library(patchwork)
library(ggsci)
library(ggrepel)
library(DESeq2)
library(readxl)
library(ggplotify)
library(plotly)
library(ggpubr)
library(tidyverse)
source('helpers.R')
```

# Introduction
Reads were first de-multiplexed with ea-utils (fastq-multx) with an allowed barcode mismatch of 1 (Default). De-multiplexed reads were then subjected to an alignment against the UniVec vector contamination database using the adapter removal tool within ea-utils (fastq-mcf). Barcodes were extracted by remove the last 12 nucleotides from each read.

Next reads were subjected to an alignment against the Oligo library using a DNA aligner. We used Bowtie2...

## Data import & preprocessing
We are going to first import the merged counts tables that is outputted from the Nextflow pipeline. This table already has the raw counts summarised from the alignments and merged into a single matrix.

## Import data
```{r}
# Import CD4 counts
counts = readRDS('../data/CD4-counts.rda')
head(counts)

# Import CD4 metadata
metadata = readRDS('../data/CD4-metadata.rda')
head(metadata)

# Import MultiQC results
report = readRDS('../data/CD4-multiqc.rda')
head(report)

# Import oligo annotations
annotations = readRDS('../data/CD4-annotations.rda')
head(annotations)
```


----

## Mapping quality
```{r}
# Mean
mean(report$raw_total_sequences)
median(report$raw_total_sequences)
range(report$raw_total_sequences)
```

### Number of mapped reads

```{r}
p1 = report %>% 
  select(Sample, reads_unmapped, reads_mapped) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped", "reads_mapped"), names_to = "feature", values_to = "reads") %>% 
    mutate(feature = if_else(feature == "reads_unmapped", "Unmapped", feature)) %>% 
  mutate(feature = if_else(feature == "reads_mapped", "Mapped", feature)) %>% 
  ggplot(aes(x = reorder(Sample, reads), y = log10(reads), fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_classic(base_size = 10) +
  theme(legend.position = "top") +
  scale_fill_brewer(palette = "Set2") +
  xlab("") +
  ylab("Number of reads (log10)") +
  ggtitle("CD4: mapping rate")
p1

### Percent of mapped reads
p2 = report %>% 
  select(Sample, reads_unmapped_percent, reads_mapped_percent ) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped_percent", "reads_mapped_percent"), names_to = "feature", values_to = "reads") %>% 
  mutate(reads = reads / 100) %>% 
  mutate(feature = if_else(feature == "reads_unmapped_percent", "Unmapped", feature)) %>% 
  mutate(feature = if_else(feature == "reads_mapped_percent", "Mapped", feature)) %>% 
  ggplot(aes(x = reorder(Sample, reads), y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_classic(base_size = 10) +
  theme(legend.position = "top") +
  scale_fill_brewer(palette = "Set2") +
  scale_y_continuous(labels = scales::percent) +
  xlab("") +
  ylab("Percent of reads")
p2

# Plot together
p = (p1 | p2) + plot_layout(guides = "collect") & theme(legend.position = "bottom")
p
ggsave("figures/Benchmark-CD4-mapping-qc.pdf", p, width = 8, height = 5)

```


----

## Normalize data

```{r}
# Match the sample ID's between samples
counts = counts[,match(metadata$Sample, colnames(counts))]

# Make DESeq2 object
ddsMat <- DESeqDataSetFromMatrix(countData = counts,
                                 colData = metadata,
                                 design = ~Group)

# Get normalize counts
ddsMat <- estimateSizeFactors(ddsMat)

# Filter flow counts
ddsMat.fil = ddsMat %>% 
  filter_low(percentile = 0.05)

# Get normalized counts 
counts.norm.fil <- (counts(ddsMat.fil, normalized = TRUE) + 0.5) %>% 
  as.data.frame() %>% 
  rownames_to_column("oligo_id")

# Create a long table 
counts.norm.fil.long = counts.norm.fil %>% 
  pivot_longer(cols = -oligo_id, names_to = "sampleId", values_to = "norm")
```

---

## Mock replicates
```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_mock_1), y = log10(cd4_mock_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("Mock replicate #1 (log10)") +
  ylab("Mock replicate #2 (log10)") +
  theme_light() +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE)
```
### SNORD73 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_snord73_1), y = log10(cd4_snord73_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("SNORD73 replicate #1 (log10)") +
  ylab("SNORD73 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE)

counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_snord73_2), y = log10(cd4_snord73_3))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("SNORD73 replicate #1 (log10)") +
  ylab("SNORD73 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 

counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_snord73_1), y = log10(cd4_snord73_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("SNORD73 replicate #1 (log10)") +
  ylab("SNORD73 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE)
```


### MANSC1 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_mansc1_1), y = log10(cd4_mansc1_2))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("MANSC1 replicate #1 (log10)") +
  ylab("MANSC1 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 

counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_mansc1_2), y = log10(cd4_mansc1_3))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("MANSC1 replicate #2 (log10)") +
  ylab("MANSC1 replicate #3 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 

counts.norm.fil %>% 
  ggplot(aes(x = log10(cd4_mansc1_1), y = log10(cd4_mansc1_3))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("MANSC1 replicate #1 (log10)") +
  ylab("MANSC1 replicate #3 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) +
  geom_smooth(method = "lm", se = FALSE) 
```

### Plot normalized histogram
```{r}
counts.norm.fil.long %>% 
  ggplot(aes(x = log10(norm))) +
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  facet_wrap(sampleId~.) +
  theme_light(base_size = 11) +
  ylab("Density") +
  xlab("Normalized reads per oligo (log10)") 
```


----

## Determine dropout oligo's

### SNORD73 vs. Mock
```{r}
snord_mock.res = compute_stats(ddsMat.fil, trt = "SNORD", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Significant oligos
snord_mock.res %>% 
  filter(padj < 0.05)

# Browse
snord_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
snord_mock.maplot = snord_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = gene_name)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(snord_mock.res, gene_name %in% c("SNORDmut")), size = 3, colour = "#efc383") +
  geom_point(data = filter(snord_mock.res, gene_name %in% c("SNORDwt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(snord_mock.res, gene_name %in% c("SNORDmut", "SNORDwt")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(SNORD73/Mock)")  +
  ggtitle("SNORD")
ggplotly(snord_mock.maplot)
```

----

### MANSC1 vs. Mock

```{r}
mansc_mock.res = compute_stats(ddsMat.fil, trt = "MANSC", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Significant oligos
mansc_mock.res %>% 
  filter(padj < 0.05) 

# Browse
mansc_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
mansc_mock.maplot = mansc_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = gene_name)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(mansc_mock.res, gene_name %in% c("MANSC1mut")), size = 3, colour = "#aec3a0") +
  geom_point(data = filter(mansc_mock.res, gene_name %in% c("MANSC1wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(mansc_mock.res, gene_name %in% c("MANSC1mut", "MANSC1wt")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(MANSC1/Mock)")  +
  ggtitle("MANSC1")
ggplotly(mansc_mock.maplot)
```

### Plot together
```{r}
p = (snord_mock.maplot + ggtitle("") | mansc_mock.maplot + ggtitle("")) & ylim(-3, 0.75) & xlim(7, 11)
p
ggsave("figures/CD4-benchmark-older.pdf", p, width = 6, height = 4)
```


### MANSC1 vs. SNORD73
```{r, eval = F}
mansc_snord.res = compute_stats(ddsMat.fil, trt = "MANSC", ref = "SNORD", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Get significant hits
mansc_snord.res %>%  
  filter(padj < 0.05) 
mansc1_snord.res

# Browse
mansc_snord.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
mansc_snord.maplot = mansc_snord.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = gene_name)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(mansc_snord.res, gene_name %in% c("MANSC1mut")), size = 3, colour = "#aec3a0") +
  geom_point(data = filter(mansc_snord.res, gene_name %in% c("MANSC1wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(mansc_snord.res, gene_name %in% c("MANSC1mut", "MANSC1wt")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(MANSC1/Mock)") 
ggplotly(mansc_snord.maplot)
```


## Session info
```{r}
sessionInfo()
```

