---
title: "CD8-Benchmark"
author: "Thomas W. Battaglia"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(fig.path = "images/")
library(knitr)
library(patchwork)
library(ggsci)
library(ggrepel)
library(DESeq2)
library(readxl)
library(ggplotify)
library(plotly)
library(ggpubr)
library(tidyverse)
source('helpers.R')
```

# Introduction
Reads were first de-multiplexed with ea-utils (fastq-multx) with an allowed barcode mismatch of 1 (Default). De-multiplexed reads were then subjected to an alignment against the UniVec vector contamination database using the adapter removal tool within ea-utils (fastq-mcf). Barcodes were extracted by remove the last 12 nucleotides from each read.

Next reads were subjected to an alignment against the Oligo library using a DNA aligner. We used Bowtie2...

## Data import & preprocessing
We are going to first import the merged counts tables that is outputted from the Nextflow pipeline. This table already has the raw counts summarised from the alignments and merged into a single matrix.

## Import data
```{r}
# Import CD4 counts
counts = readRDS('../data/CD8-counts.rda')
head(counts)

# Import CD4 metadata
metadata = readRDS('../data/CD8-metadata.rda')
head(metadata)

# Import MultiQC results
report = readRDS('../data/CD8-multiqc.rda')
head(report)

# Import oligo annotations
annotations = readRDS('../data/CD8-annotations.rda') 
head(annotations)
```


----

## Mapping quality
```{r}
# Mean
mean(report$raw_total_sequences)
median(report$raw_total_sequences)
range(report$raw_total_sequences)
```

### Number of mapped reads

```{r}
p1 = report %>% 
  select(Sample, reads_unmapped, reads_mapped) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped", "reads_mapped"), names_to = "feature", values_to = "reads") %>% 
    mutate(feature = if_else(feature == "reads_unmapped", "Unmapped", feature)) %>% 
  mutate(feature = if_else(feature == "reads_mapped", "Mapped", feature)) %>% 
  ggplot(aes(x = reorder(Sample, reads), y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_classic(base_size = 10) +
  theme(legend.position = "top") +
  scale_fill_brewer(palette = "Set2") +
  xlab("") +
  ylab("Number of reads") +
  ggtitle("CD8: mapping rate")
p1

### Percent of mapped reads
p2 = report %>% 
  select(Sample, reads_unmapped_percent, reads_mapped_percent ) %>% 
  mutate(Sample = str_replace(Sample, '-bamstats', '')) %>% 
  pivot_longer(cols = c("reads_unmapped_percent", "reads_mapped_percent"), names_to = "feature", values_to = "reads") %>% 
  mutate(reads = reads / 100) %>% 
  mutate(feature = if_else(feature == "reads_unmapped_percent", "Unmapped", feature)) %>% 
  mutate(feature = if_else(feature == "reads_mapped_percent", "Mapped", feature)) %>% 
  ggplot(aes(x = reorder(Sample, reads), y = reads, fill = feature)) +
  geom_col() +
  coord_flip() +
  theme_classic(base_size = 10) +
  theme(legend.position = "top") +
  scale_fill_brewer(palette = "Set2") +
  scale_y_continuous(labels = scales::percent) +
  xlab("") +
  ylab("Percent of reads")
p2

# Plot together
p = (p1 | p2) + plot_layout(guides = "collect") & theme(legend.position = "bottom")
p
ggsave("figures/Benchmark-CD8-mapping-qc.pdf", p, width = 8, height = 5)

```


----

## Normalize data

```{r}
# Match the sample ID's between samples
counts = counts[,match(metadata$Sample, colnames(counts))]

# Make DESeq2 object
ddsMat <- DESeqDataSetFromMatrix(countData = counts,
                                 colData = metadata,
                                 design = ~Group)

# Get normalize counts
ddsMat <- estimateSizeFactors(ddsMat)

# Filter flow counts
ddsMat.fil = ddsMat %>% 
  filter_low(percentile = 0.05)

# Get normalized counts 
counts.norm.fil <- (counts(ddsMat.fil, normalized = TRUE) + 0.5) %>% 
  as.data.frame() %>% 
  rownames_to_column("oligo_id")

# Create a long table 
counts.norm.fil.long = counts.norm.fil %>% 
  pivot_longer(cols = -oligo_id, names_to = "sampleId", values_to = "norm")
```

---

## Mock replicates

```{r}
counts.norm.fil %>% 
  ggplot(., aes(x = log10(mock_a), y = log10(mock_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("Mock replicate #1 (log10)") +
  ylab("Mock replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```

### 1D3 replicates
```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(id3_a), y = log10(id3_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1D3 replicate #1 (log10)") +
  ylab("1D3 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```


### CDK4-53 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(cdk453_a), y = log10(cdk453_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("CDK4/53 replicate #1 (log10)") +
  ylab("CDK4/53 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```


### Dilution 1:10 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(d1_10_a), y = log10(d1_10_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:10 replicate #1 (log10)") +
  ylab("1:10 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..))
```

### Dilution 1:100 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(d1_100_a), y = log10(d1_100_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:100 replicate #1 (log10)") +
  ylab("1:100 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```

### Dilution 1:300 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(d1_300_a), y = log10(d1_300_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:300 replicate #1 (log10)") +
  ylab("1:300 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```


### Dilution 1:1000 replicates

```{r}
counts.norm.fil %>% 
  ggplot(aes(x = log10(d1_1000_a), y = log10(d1_1000_b))) +
  geom_point(colour = alpha("grey", 0.7)) +
  xlab("1:1000 replicate #1 (log10)") +
  ylab("1:1000 replicate #2 (log10)") +
  theme_light(base_size = 11) +
  stat_cor(method = "pearson", aes(label = ..r.label..)) 
```

### Plot normalized histogram
```{r}
counts.norm.fil.long %>% 
  ggplot(aes(x = log10(norm))) +
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  geom_density(alpha = 0.2, fill = "#FF6666") +
  facet_wrap(sampleId~.) +
  theme_light(base_size = 11) +
  ylab("Density") +
  xlab("Normalized reads per oligo (log10)") 
```


----

## Determine dropout oligo's

### MART1/1D3 vs. Mock
```{r}
mart1_mock.res = compute_stats(ddsMat.fil, trt = "ID3", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) 

# Significant oligos
mart1_mock.res %>% 
  filter(padj < 0.05)

# Browse
mart1_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
mart1_mock.maplot = mart1_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = oligo_id)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(mart1_mock.res, gene_name %in% c("MART1_ELA")), size = 3, colour = "#efc383") +
  geom_point(data = filter(mart1_mock.res, gene_name %in% c("gene812")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(mart1_mock.res, gene_name %in% c("CDK4mut")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(mart1_mock.res, gene_name %in% c("CDK4wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(mart1_mock.res, oligo_id %in% c("MART1_ELA", "MART1", "CDK4mut", "CDK4wt", "gene812")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(MART1/Mock)")  +
  ggtitle("MART1")
ggplotly(mart1_mock.maplot)
```


### D10
```{r}
d10_mock.res = compute_stats(ddsMat.fil, trt = "D10", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Significant oligos
d10_mock.res %>% 
  filter(padj < 0.05)

# Browse
d10_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
d10_mock.maplot = d10_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = oligo_id)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(d10_mock.res, gene_name %in% c("MART1_ELA")), size = 3, colour = "#efc383") +
  geom_point(data = filter(d10_mock.res, gene_name %in% c("gene812")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d10_mock.res, gene_name %in% c("CDK4mut")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d10_mock.res, gene_name %in% c("CDK4wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(d10_mock.res, oligo_id %in% c("MART1_ELA", "MART1", "CDK4mut", "CDK4wt", "gene812")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(D10/Mock)") 
ggplotly(d10_mock.maplot)
```


### D100
```{r}
d100_mock.res = compute_stats(ddsMat.fil, trt = "D100", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Significant oligos
d100_mock.res %>% 
  filter(padj < 0.05) 

# Browse
d100_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
d100_mock.maplot = d100_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = oligo_id)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(d100_mock.res, gene_name %in% c("MART1_ELA")), size = 3, colour = "#efc383") +
  geom_point(data = filter(d100_mock.res, gene_name %in% c("gene812")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d100_mock.res, gene_name %in% c("CDK4mut")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d100_mock.res, gene_name %in% c("CDK4wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(d100_mock.res, oligo_id %in% c("MART1_ELA", "MART1", "CDK4mut", "CDK4wt", "gene812")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(D100/Mock)") 
ggplotly(d100_mock.maplot)
```

### D300
```{r}
d300_mock.res = compute_stats(ddsMat.fil, trt = "D300", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Significant oligos
d300_mock.res %>% 
  filter(padj < 0.05) 

# Browse
d300_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
d300_mock.maplot = d300_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = oligo_id)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(d300_mock.res, gene_name %in% c("MART1_ELA")), size = 3, colour = "#efc383") +
  geom_point(data = filter(d300_mock.res, gene_name %in% c("gene812")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d300_mock.res, gene_name %in% c("CDK4mut")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d300_mock.res, gene_name %in% c("CDK4wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(d300_mock.res, oligo_id %in% c("MART1_ELA", "MART1", "CDK4mut", "CDK4wt", "gene812")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(D300/Mock)") 
ggplotly(d300_mock.maplot)
```

### D1000
```{r}
d1000_mock.res = compute_stats(ddsMat.fil, trt = "D1000", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Amut", "SNORDmut", gene_name)) %>% 
  mutate(gene_name = if_else(gene_name == "RPS3Awt", "SNORDwt", gene_name))

# Significant oligos
d1000_mock.res %>% 
  filter(padj < 0.05) 

# Browse
d1000_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
d1000_mock.maplot = d1000_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = oligo_id)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(data = filter(d1000_mock.res, gene_name %in% c("MART1_ELA")), size = 3, colour = "#efc383") +
  geom_point(data = filter(d1000_mock.res, gene_name %in% c("gene812")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d1000_mock.res, gene_name %in% c("CDK4mut")), size = 3, colour = "#a6a6a6") +
  geom_point(data = filter(d1000_mock.res, gene_name %in% c("CDK4wt")), size = 3, colour = "#a6a6a6") +
  geom_text_repel(data = filter(d1000_mock.res, oligo_id %in% c("MART1_ELA", "MART1", "CDK4mut", "CDK4wt", "gene812")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(D1000/Mock)") 
ggplotly(d1000_mock.maplot)
```



-----

### CDK4#53 vs. Mock

```{r}
cdk454_mock.res = compute_stats(ddsMat.fil, trt = "CDK4", ref = "Mock", fitType = "parametric", lfcThreshold = 0.25, alpha = 0.05) %>% 
  left_join(annotations) %>% 
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>% 
  mutate(log2FoldChangeShrunken = round(log2FoldChangeShrunken, 2)) 

# Significant oligos
cdk454_mock.res %>% 
  filter(padj < 0.05)

# Browse
cdk454_mock.res %>% 
  select(gene_id, log2FoldChange, padj, gene_name, gene_id) %>% 
  DT::datatable(extensions = 'Buttons', options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
  ))

# Plot MA-plot
cdk453_mock.maplot = cdk454_mock.res %>% 
  ggplot(aes(x = log2(baseMean), y = log2FoldChange, label = oligo_id)) +
  geom_point(colour = alpha("grey", 0.7)) +
  geom_point(colour = alpha("red", 0.7), data = . %>% filter(oligo_id %in% c("MART1_ELA_a", "MART1_ELA_b"))) +
  geom_point(colour = alpha("lightblue", 0.7), data = . %>% filter(oligo_id %in% c("MART1_a", "MART1_b"))) +
  geom_point(colour = alpha("orange", 0.7), data = . %>% filter(oligo_id %in% c("CDK4wt_a", "CDK4wt_b"))) +
  geom_point(colour = alpha("darkgreen", 0.7), data = . %>% filter(oligo_id %in% c("CDK4mut_a", "CDK4mut_b"))) +
  geom_text_repel(data = filter(cdk454_mock.res, oligo_id %in% c("CDK4wt_a", "CDK4wt_b", "CDK4mut_a", "CDK4mut_b")), box.padding = 0.5, max.overlaps = Inf, size = 2.5) +
  theme_classic() +
  theme_classic() +
  xlab("Mean counts (log2)") +
  ylab("Fold change log2(CDK4 #53/Mock)")  +
  ggtitle("CDK4 #53")
ggplotly(cdk453_mock.maplot)
```


----

## Session info
```{r}
sessionInfo()
```

